### PR TITLE
Feature: 网关规则和自定义api规则动态修改增加对写数据源的支持

### LINKED ISSUE #992: Support registering writable data-source for GatewayFlowRule and customized ApiDefinition
WritableDataSourceRegistry 和 GatewayRuleManager不支持GatewayFlowRule注册写数据源

### PR DESCRIPTION
**Pull request by zhangyide**
### Does this pull request fix one issue?

Resolves #992

### CODE REVIEW

**Comment by sczyh30** [comment]
Could you please sign the CLA here: https://cla-assistant.io/alibaba/Sentinel?pullRequest=1057

**Comment by zhangyide** [comment]
> Could you please sign the CLA here: https://cla-assistant.io/alibaba/Sentinel?pullRequest=1057

ok

File: sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java [hunk 1]
```diff
@@ -54,8 +57,41 @@
         String result = SUCCESS_MSG;
         List<GatewayFlowRule> flowRules = JSONArray.parseArray(data, GatewayFlowRule.class);
         GatewayRuleManager.loadRules(new HashSet<>(flowRules));
+        if (!writeToDataSource(gatewayFlowWds, new HashSet<>(flowRules))) {
```
Comment (cdfive) [inline sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: Here creating HashSet twice, how about optimizing it?

Comment (zhangyide) [reply sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: ok，i will optimize it.

Comment (zhangyide) [reply sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: @cdfive I had optimized it.

File: sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java [hunk 2]
```diff
@@ -52,10 +54,42 @@
         RecordLog.info(String.format("[API Server] Receiving rule change (type: gateway rule): %s", data));
 
         String result = SUCCESS_MSG;
-        List<GatewayFlowRule> flowRules = JSONArray.parseArray(data, GatewayFlowRule.class);
-        GatewayRuleManager.loadRules(new HashSet<>(flowRules));
+        Set<GatewayFlowRule> flowRules = JSON.parseObject(data,new TypeReference<Set<GatewayFlowRule>>(){});
```
Comment (sczyh30) [inline sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: Could you please reformat your code?

Comment (cdfive) [reply sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: @zhangyide One space after comma makes code better reading, using IDE to format this line is okay : )

Comment (zhangyide) [reply sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/command/UpdateGatewayRuleCommandHandler.java]: ok. I had reformated it.

**Comment by sczyh30** [comment]
Thanks for contributing!

**Pull request merged**